# Predicting the rating of a dress from online reviews

In [30]:
import seaborn as sns
import sklearn as sk
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

## Data set

In [2]:
df = pd.read_csv("clothingreviews.csv")
df.head(30) 

Unnamed: 0  Clothing ID  Age  \
0            0          767   33   
1            1         1080   34   
2            2         1077   60   
3            3         1049   50   
4            4          847   47   
5            5         1080   49   
6            6          858   39   
7            7          858   39   
8            8         1077   24   
9            9         1077   34   
10          10         1077   53   
11          11         1095   39   
12          12         1095   53   
13          13          767   44   
14          14         1077   50   
15          15         1065   47   
16          16         1065   34   
17          17          853   41   
18          18         1120   32   
19          19         1077   47   
20          20          847   33   
21          21         1080   55   
22          22         1077   31   
23          23         1077   34   
24          24          847   55   
25          25          697   31   
26          26          949   33   
27          27         1003   31   
28          28          684   53   
29          29            4   28   

                                                Title  \
0                                                 NaN   
1                                                 NaN   
2                             Some major design flaws   
3                                    My favorite buy!   
4                                    Flattering shirt   
5                             Not for the very petite   
6                                Cagrcoal shimmer fun   
7                Shimmer, surprisingly goes with lots   
8                                          Flattering   
9                                   Such a fun dress!   
10       Dress looks like it's made of cheap material   
11                                                NaN   
12                                         Perfect!!!   
13                                           Runs big   
14                Pretty party dress with some issues   
15                          Nice, but not for my body   
16  You need to be at least average height, or taller   
17                       Looks great with white pants   
18                                Super cute and cozy   
19                            Stylish and comfortable   
20                                  Cute, crisp shirt   
21                                          I'm torn!   
22                             Not what it looks like   
23                        Like it, but don't love it.   
24                                          Versatile   
25                                         Falls flat   
26                                Huge disappointment   
27                                Loved, but returned   
28                                     Great shirt!!!   
29                               Great layering piece   

                                          Review Text  Rating  \
0   Absolutely wonderful - silky and sexy and comf...       4   
1   Love this dress!  it's sooo pretty.  i happene...       5   
2   I had such high hopes for this dress and reall...       3   
3   I love, love, love this jumpsuit. it's fun, fl...       5   
4   This shirt is very flattering to all due to th...       5   
5   I love tracy reese dresses, but this one is no...       2   
6   I aded this in my basket at hte last mintue to...       5   
7   I ordered this in carbon for store pick up, an...       4   
8   I love this dress. i usually get an xs but it ...       5   
9   I'm 5"5' and 125 lbs. i ordered the s petite t...       5   
10  Dress runs small esp where the zipper area run...       3   
11  This dress is perfection! so pretty and flatte...       5   
12  More and more i find myself reliant on the rev...       5   
13  Bought the black xs to go under the larkspur m...       5   
14  This is a nice choice for holiday gatherings. ...       3   
15  I took these out of the package and wanted the...       4   
16  Material a

To make sure I have only dresses in my dataframe I'm going to subset the dataframe.

In [7]:
df_subset = df[df["Department Name"] == "Dresses"]
df_subset.head(30)

Unnamed: 0  Clothing ID  Age  \
1             1         1080   34   
2             2         1077   60   
5             5         1080   49   
8             8         1077   24   
9             9         1077   34   
10           10         1077   53   
11           11         1095   39   
12           12         1095   53   
14           14         1077   50   
19           19         1077   47   
21           21         1080   55   
22           22         1077   31   
23           23         1077   34   
52           52         1104   39   
58           58         1095   36   
60           60         1095   41   
69           69         1078   56   
74           74         1081   59   
76           76         1082   31   
79           79         1081   62   
80           80         1072   68   
87           87         1081   39   
90           90         1078   51   
93           93         1081   31   
109         109         1081   47   
111         111         1081   44   
113         113         1081   61   
114         114         1082   32   
116         116         1072   51   
117         117         1078   32   

                                            Title  \
1                                             NaN   
2                         Some major design flaws   
5                         Not for the very petite   
8                                      Flattering   
9                               Such a fun dress!   
10   Dress looks like it's made of cheap material   
11                                            NaN   
12                                     Perfect!!!   
14            Pretty party dress with some issues   
19                        Stylish and comfortable   
21                                      I'm torn!   
22                         Not what it looks like   
23                    Like it, but don't love it.   
52                                            NaN   
58                                            NaN   
60                                Gorgeous dress!   
69                           Great summer fabric!   
74                                            NaN   
76                    Runs large in top and waist   
79                      So cute...but dry clean?!   
80                          Elegant classic dress   
87                          True to size to large   
90                         Sweet flattering dress   
93                                            NaN   
109                                           NaN   
111                           An almost for me...   
113                                Cute and comfy   
114                         Disappointing quality   
116                             Very vintage feel   
117                                           NaN   

                                           Review Text  Rating  \
1    Love this dress!  it's sooo pretty.  i happene...       5   
2    I had such high hopes for this dress and reall...       3   
5    I love tracy reese dresses, but this one is no...       2   
8    I love this dress. i usually get an xs but it ...       5   
9    I'm 5"5' and 125 lbs. i ordered the s petite t...       5   
10   Dress runs small esp where the zipper area run...       3   
11   This dress is perfection! so pretty and flatte...       5   
12   More and more i find myself reliant on the rev...       5   
14   This is a nice choice for holiday gatherings. ...       3   
19   I love the look and feel of this tulle dress. ...       5   
21   I'm upset because for the price of the dress, ...       4   
22   First of all, this is not pullover styling. th...       2   
23   Cute little dress fits tts. it is a little hig...       3   
52   Love the color and style, but material snags e...       3   
58   I got this in the petite length, size o, and i...       4   
60   This dress is simply beautiful and stunning. i...       5   
69   I really wanted this to work. alas, it had a s...       3   
74   I passed up this dre

## Text pre-processing steps resulting in a document-feature matrix

In [9]:
text = df["Review Text"].values.astype("U") #Taking the text from the df. We need to convert it to Unicode
vect = CountVectorizer(stop_words= "english") #Create the CV object, with English stop words
vect = vect.fit(text) #We fit the model with the words from the review text
feature_names = vect.get_feature_names() #Get the words from the vocabulary

In [10]:
docu_feat = vect.transform(text) # make a matrix

In [11]:
print(docu_feat[0:500,0:500])

  (5, 10)	1
  (9, 98)	1
  (9, 354)	1
  (13, 289)	1
  (13, 297)	1
  (13, 352)	1
  (13, 371)	1
  (16, 16)	1
  (18, 133)	1
  (20, 122)	1
  (25, 310)	1
  (25, 377)	1
  (31, 150)	1
  (33, 398)	2
  (37, 118)	1
  (37, 133)	1
  (37, 310)	1
  (37, 474)	1
  (43, 98)	1
  (51, 146)	1
  (55, 377)	1
  (70, 424)	1
  (70, 496)	1
  (72, 355)	1
  (75, 49)	1
  :	:
  (422, 73)	1
  (422, 348)	1
  (423, 63)	1
  (423, 277)	1
  (423, 316)	1
  (428, 49)	1
  (429, 15)	1
  (433, 15)	1
  (434, 28)	1
  (434, 353)	1
  (436, 59)	1
  (438, 63)	1
  (440, 114)	1
  (446, 360)	1
  (450, 15)	1
  (457, 103)	1
  (461, 371)	1
  (461, 496)	1
  (462, 378)	1
  (475, 150)	1
  (475, 354)	1
  (477, 95)	1
  (481, 63)	1
  (491, 118)	1
  (496, 15)	2


## Building the model

In [13]:
X = docu_feat #the document-feature matrix is the X matrix
y = df["Rating"] #creating the y vector
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3) #split the data and store it

In [17]:
nb = MultinomialNB() #create the model
nb = nb.fit(X_train, y_train) #fit the model X=features, y=character

## Model evaluation

In [18]:
nb_score = nb.score(X_test, y_test)
print(f"The accuracy for Naive Bayes: {nb_score}")

The accuracy for Naive Bayes: 0.6234743116661936


To make sure the accuracy of the model is better than if it would be random. I want to calculate the value counts of the rating.

In [21]:
df_subset["Rating"].value_counts() #Let's have a look at the 'genre' variable

5    3397
4    1395
3     838
2     461
1     228
Name: Rating, dtype: int64

In [24]:
228+461+838+1395+3397

6319

In [25]:
1395+3397

4792

In [27]:
(4792/9319)

0.5142182637622063

The accuracy of the Naive Bayes is better than if it would be random. 
NB is 0.62 whereas random is 0.51

I am a little bit confused how to do the test cases. It would be nice to have it explained one more time. 